tests/system/src/test/java/test/javafx/scene/StyleMemoryLeakTest.java

```diff
@@ -0,0 +1,106 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package test.javafx.scene;
+
+import javafx.application.Application;
+import javafx.application.Platform;
+import javafx.scene.Group;
+import javafx.scene.Node;
+import javafx.scene.Parent;
+import javafx.scene.control.Button;
+import javafx.scene.Scene;
+import javafx.stage.Stage;
+import junit.framework.Assert;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import test.util.Util;
+import test.util.memory.JMemoryBuddy;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import static org.junit.Assert.fail;
+
+
+public class StyleMemoryLeakTest {
+
+    static CountDownLatch startupLatch;
+    static Stage stage;
+
+    public static class TestApp extends Application {
+
+        @Override
+        public void start(Stage primaryStage) throws Exception {
+            stage = primaryStage;
+            startupLatch.countDown();
+        }
+    }
+
+    @BeforeClass
+    public static void initFX() {
+        startupLatch = new CountDownLatch(1);
+        new Thread(() -> Application.launch(StyleMemoryLeakTest.TestApp.class, (String[])null)).start();
+        try {
+            if (!startupLatch.await(15, TimeUnit.SECONDS)) {
+                fail("Timeout waiting for FX runtime to start");
+            }
+        } catch (InterruptedException ex) {
+            fail("Unexpected exception: " + ex);
+        }
+    }
+
+    @Test
+    public void testRootNodeMemoryLeak() throws Exception {
+        JMemoryBuddy.memoryTest((checker) -> {
+            Parent toBeRemoved = new Button();
+            Group root = new Group();
+            root.getChildren().add(toBeRemoved);
+            Util.runAndWait(() -> {
+                stage.setScene(new Scene(root));
+                stage.show();
+            });
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            Util.runAndWait(() -> {
+                root.getChildren().clear();
+                stage.close();
+            });
+            checker.assertCollectable(stage);
+            checker.setAsReferenced(toBeRemoved);
+            stage = null;
+        });
+    }
+
+    @AfterClass
+    public static void teardownOnce() {
+        Platform.runLater(() -> {
+            stage.hide();
+            Platform.exit();
+        });
+    }
+}
```

Comment:
In order to make this test similar to existing system tests, I made some relevant changes. Modified test is added below.
The modified test fails with this fix, but I expected it to pass. Can you please check this.

Changes are 
1. `Thread.sleep()` is removed.
2. `root` and `toBeRemoved` button are now class members.
3. Scenegraph is constructed and shown in `TestApp.start()` method.


```
public class StyleMemoryLeakTest {

    static CountDownLatch startupLatch;
    static Stage stage;
    static Button toBeRemoved;
    static Group root;

    public static class TestApp extends Application {

        @Override
        public void start(Stage primaryStage) throws Exception {
            stage = primaryStage;
            toBeRemoved = new Button();
            root = new Group();
            root.getChildren().add(toBeRemoved);
            stage.setScene(new Scene(root));
            stage.setOnShown(l -> {
                Platform.runLater(() -> startupLatch.countDown());
            });
            stage.show();
        }
    }

    @BeforeClass
    public static void initFX() throws Exception {
        startupLatch = new CountDownLatch(1);
        new Thread(() -> Application.launch(StyleMemoryLeakTest.TestApp.class, (String[])null)).start();
        assertTrue("Timeout waiting for FX runtime to start", startupLatch.await(15, TimeUnit.SECONDS));
    }

    @Test
    public void testRootNodeMemoryLeak() throws Exception {
        Util.runAndWait(() -> {
            root.getChildren().clear();
            stage.hide();
        });
        JMemoryBuddy.memoryTest((checker) -> {
            checker.assertCollectable(stage);
            checker.setAsReferenced(toBeRemoved);
            stage = null;
        });
    }

    @AfterClass
    public static void teardownOnce() {
        Platform.runLater(() -> {
            Platform.exit();
        });
    }
}
```

Reply:
I've added your verison of the unit-test. You forgot `root = null;` which was why the test failed.

If I would rewrite the test, I would put everything into the TestMethod. Because this way it's not necessary to set all the class-variables to null. It also wouldn't reuse the Stage of the Application. The scope of the test would be much smaller, because the actual test and the initialization of JavaFX would be separated. 

Should I change it that way?

Reply:
I am fine with that too. In that case my only suggestion would be, use method `Stage.setOnShown()` like in the current version to make sure that `Stage` is shown and not to rely on `Thread.sleep()`.




> It also wouldn't reuse the Stage of the Application.

In that case, test will not require the `TestApp` class. You can use `Platform.startup()` method to start JavaFX runtime and also **may** need a call to `Platform.setImplicitExit​(false)` to make sure JavaFX runtime does not exit when Stage is hidden. This way we can add multiple tests that create and hide stage in a same test file.